### PR TITLE
server/schedulers/balance_leader: source and target can share 1 candidatestores

### PR DESCRIPTION
**Pull request by AndreMouche**
### What problem does this PR solve?

Issue Number: Close #[5189](https://github.com/tikv/pd/issues/5189)

### Check List

Tests

- Unit test
-  bench mark test
- 
#### master
```
schedulers git:(master) ✗ go test -benchmem -run=^$ -bench ^BenchmarkBalanceLeader$ -benchtime=100x
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/schedulers
cpu: VirtualApple @ 2.50GHz
BenchmarkBalanceLeader-8             100         107248906 ns/op         3620522 B/op      52344 allocs/op
PASS
ok      github.com/tikv/pd/server/schedulers    30.150s
```

#### current branch 
```
schedulers git:(balance_leader_share_candidatestores) ✗ go test -benchmem -run=^$ -bench ^BenchmarkBalanceLeader$ -benchtime=100x
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/schedulers
cpu: VirtualApple @ 2.50GHz
BenchmarkBalanceLeader-8             100          42963967 ns/op         3135231 B/op      45180 allocs/op
PASS
ok      github.com/tikv/pd/server/schedulers    22.796s
```

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by HunDunDM — APPROVED** [review]

File: server/schedulers/balance_leader.go [hunk 1]
```diff
@@ -357,54 +369,66 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster) []*operator.
 	scoreFunc := func(store *core.StoreInfo) float64 {
 		return store.LeaderScore(plan.kind.Policy, plan.GetOpInfluence(store.GetID()))
 	}
-	sourceCandidate := newCandidateStores(filter.SelectSourceStores(stores, l.filters, cluster.GetOpts()), false, scoreFunc)
-	targetCandidate := newCandidateStores(filter.SelectTargetStores(stores, l.filters, cluster.GetOpts()), true, scoreFunc)
+	storesCandidate := newCandidateStores(filter.SelectSourceStores(stores, l.filters, cluster.GetOpts()), scoreFunc)
```
Comment (rleungx) [inline server/schedulers/balance_leader.go:372]: There is a little difference between select source and target when filtering.

Comment (AndreMouche) [reply server/schedulers/balance_leader.go:372]: ono

Comment (AndreMouche) [reply server/schedulers/balance_leader.go:372]: > There is a little difference between select source and target when filtering.

This PR should be closed for the previous reason.

**Comment by AndreMouche** [comment]
Since the source

**Pull request closed without merging**